File: Java-and-C++/AnalyticFunctions/Rank.cpp
/* Copyright (c) 2005 - 2015 Hewlett Packard Enterprise Development LP  -*- C++ -*-*/
/* 
 * Description: Example User Defined Analytic Function: Output nth row
 *
 * Create Date: Nov 22, 2011
 */
#include "Vertica.h"
#include <sstream>

using namespace Vertica;
using namespace std;



/**
 * User defined analytic function: Rank - works mostly the same as SQL-99 rank
 * with the ability to define as many order by columns as desired
 *
 * One difference is UDAn's do not currently require an order by clause in the DDL
 */
class Rank : public AnalyticFunction
{
    virtual void processPartition(ServerInterface &srvInterface, 
                                  AnalyticPartitionReader &inputReader, 
                                  AnalyticPartitionWriter &outputWriter)
    {
        try {
            rank = 0;
            numRowsWithSameOrder = 1;
            do {
                if (!inputReader.isNewOrderByKey()) {
                    ++numRowsWithSameOrder;
                }
                else {
                    rank += numRowsWithSameOrder;
                    numRowsWithSameOrder = 1;
                }
                outputWriter.setInt(0, rank);
                outputWriter.next();
            } while (inputReader.next());
        } catch(exception& e) {
            // Standard exception. Quit.
            vt_report_error(0, "Exception while processing partition: %s", e.what());
        }
    }

private:
    vint rank, numRowsWithSameOrder;
};

class RankFactory : public AnalyticFunctionFactory
{
    virtual void getPrototype(ServerInterface &srvInterface, ColumnTypes &argTypes, ColumnTypes &returnType)
    {
        returnType.addInt();
    }

    virtual void getReturnType(ServerInterface &srvInterface, 
                               const SizedColumnTypes &inputTypes, 
                               SizedColumnTypes &outputTypes)
    {
        outputTypes.addInt();
    }

    virtual void getParameterType(ServerInterface &srvInterface,
                                  SizedColumnTypes &parameterTypes)
    {
        // No parameters needed for Analytic Function Rank
    }

    virtual AnalyticFunction *createAnalyticFunction(ServerInterface &srvInterface)
    { return vt_createFuncObject<Rank>(srvInterface.allocator); }

};

RegisterFactory(RankFactory);

File: Java-and-C++/JavaUDx/UDAnalytics/com/vertica/JavaLibs/Rank.java
/* Copyright (c) 2005 - 2016 Hewlett Packard Enterprise Development LP  -*- Java -*-*/
/* 
 * Description: Example User Defined Analytic Function: Output nth row
 *
 * Create Date: Jan 7, 2015
 */

package com.vertica.JavaLibs;

import com.vertica.sdk.AnalyticFunction;
import com.vertica.sdk.AnalyticPartitionReader;
import com.vertica.sdk.AnalyticPartitionWriter;
import com.vertica.sdk.DestroyInvocation;
import com.vertica.sdk.ServerInterface;
import com.vertica.sdk.UdfException;

/**
 * User defined analytic function: Rank - works mostly the same as SQL-99 rank
 * with the ability to define as many order by columns as desired
 *
 * One difference is UDAn's do not currently require an order by clause in the DDL
 */
public class Rank extends AnalyticFunction {
	private int rank, numRowsWithSameOrder;

	@Override
	public void processPartition(ServerInterface srvInterface,
			AnalyticPartitionReader inputReader, AnalyticPartitionWriter outputWriter)
			throws UdfException, DestroyInvocation {
		rank = 0;
		numRowsWithSameOrder = 1;
		do{
			if(!inputReader.isNewOrderByKey()){
				++numRowsWithSameOrder;
			}else {
				rank += numRowsWithSameOrder;
				numRowsWithSameOrder = 1;
			}
			outputWriter.setLong(0, rank);
			outputWriter.next();
		}while(inputReader.next());
	}

}
